packages/metals-vscode/src/hotCodeReplace.ts

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+// Adapter from https://github.com/microsoft/vscode-java-debug/blob/main/src/hotCodeReplace.ts
+
+import * as vscode from "vscode";
+
+import { DebugSession, commands } from "vscode";
+
+export function initializeHotCodeReplace() {
+  vscode.workspace.onDidChangeConfiguration((event) => {
+    if (event.affectsConfiguration("metals.debug.settings.hotCodeReplace")) {
+      vscode.commands.executeCommand(
+        "setContext",
+        "scalaHotReloadOn",
+        hotReplaceIsOn()
+      );
+    }
+  });
+  vscode.debug.onDidStartDebugSession((session) => {
+    if (session?.configuration.noDebug && !vscode.debug.activeDebugSession) {
+      vscode.commands.executeCommand("setContext", "scalaHotReloadOn", false);
+    }
+  });
+  vscode.debug.onDidChangeActiveDebugSession((session) => {
+    vscode.commands.executeCommand(
+      "setContext",
+      "scalaHotReloadOn",
+      session && !session.configuration.noDebug && hotReplaceIsOn()
+    );
+  });
+}
+
+export async function applyHCR() {
+  const debugSession: DebugSession | undefined =
+    vscode.debug.activeDebugSession;
+  if (!debugSession) {
+    return;
+  }
+
+  const serverVersion =
+    vscode.workspace.getConfiguration("metals").get<string>("serverVersion") ??
+    "";
+  if (serverVersion < "1.0.1") {
```

Comment:
I think we could just make sure we update DAP in Metals and in Bloop first and once that is available we can merge this PR and update the version to the most recent snapshot at the time. If anyone switches to an older version of Metals it's fine if we fail.